compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -144,59 +164,63 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         for (_, error_set) in error_map.iter() {
             // We want to suppress "duplicate" errors with the same span.
             for error in error_set {
-                if let Some(index) = error.index {
+                let Some((index, error_source)) = error.source else {
+                    continue;
+                };
+
+                for error2 in error_set {
                     // Suppress errors that are either:
                     // 1) strictly implied by another error.
                     // 2) implied by an error with a smaller index.
-                    for error2 in error_set {
-                        if error2.index.is_some_and(|index2| is_suppressed[index2]) {
-                            // Avoid errors being suppressed by already-suppressed
-                            // errors, to prevent all errors from being suppressed
-                            // at once.
-                            continue;
-                        }
+                    if self.error_implied_by(error.predicate, error2.predicate)
+                        && (!error2.source.is_some_and(|(index2, _)| index2 >= index)
+                            || !self.error_implied_by(error2.predicate, error.predicate))
+                    {
+                        info!("skipping `{}` (implied by `{}`)", error.predicate, error2.predicate);
+                        is_suppressed[index] = true;
+                        break;
+                    }
 
-                        if self.error_implies(error2.predicate, error.predicate)
-                            && !(error2.index >= error.index
-                                && self.error_implies(error.predicate, error2.predicate))
-                        {
-                            info!("skipping {:?} (implied by {:?})", error, error2);
-                            is_suppressed[index] = true;
-                            break;
-                        }
+                    // Also suppress the error if we are absolutely certain that a different
+                    // error is the one that the user should fix. This will suppress errors
+                    // about `<T as Pointee>::Metadata == ()` that can be fixed by `T: Sized`.
+                    if error.predicate.to_opt_poly_projection_pred().is_some()
+                        && error2.predicate.to_opt_poly_trait_pred().is_some()
+                        && self.error_fixed_by(
```

Comment:
I wonder if we can just use `error_fixed_by` everywhere -- it seems more general than `error_implied_by`?

Reply:
Not necessarily in this PR, but maybe tracked as a FIXME.

Reply:
That's what I tried first, but there are multiple problems with this approach:
* `error_fixed_by` does not work when there are any infer vars in `cond` (first arg). This adds a lot of duplicate diagnostics.
* There are some cases where `A fixed-by B` + `not (B fixed-by A)` + `B fixed-by C` + `not (C fixed-by B)` + `C fixed-by A` + `not (A fixed-by C)`. In these cases, replacing `error_implied_by` with `error_fixed_by` causes no diagnostics to be emitted. This cannot happen with `error_implied_by`, because `A implied-by B` is transitive.

Reply:
> * There are some cases where `A fixed-by B` + `not (B fixed-by A)` + `B fixed-by C` + `not (C fixed-by B)` + `C fixed-by A` + `not (A fixed-by C)`.

After thinking about this again, I'm actually not sure anymore whether this can actually happen and can't come up with a concrete example. I had a test failing and thought that was due a cycle like that, but after trying it again it turned out to be an unrelated implementation bug.